skmonitor/src/Monitor.sk

```diff
@@ -0,0 +1,398 @@
+    };
+    SKStore.restoreFromTrace()
+  })
+}
+
+fun setBoolAttribute(name: String, value: Bool): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setBoolAttribute(SKStore.duplicate(name), value);
+    SKStore.restoreFromTrace()
+  })
+}
+
+fun setStringArrayAttribute(name: String, value: Array<String>): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setArrayAttribute(
+      SKStore.duplicate(name),
+      value.map(v ~> VString(SKStore.duplicate(v))),
+    );
+    SKStore.restoreFromTrace()
+  })
+}
+
+fun setIntArrayAttribute(name: String, value: Array<Int>): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setArrayAttribute(SKStore.duplicate(name), value.map(v ~> VInt(v)));
+    SKStore.restoreFromTrace()
+  })
+}
+
+fun setFloatArrayAttribute(name: String, value: Array<Float>): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setArrayAttribute(SKStore.duplicate(name), value.map(v ~> VFloat(v)));
+    SKStore.restoreFromTrace()
+  })
+}
+
+fun setBoolArrayAttribute(name: String, value: Array<Bool>): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setArrayAttribute(SKStore.duplicate(name), value.map(v ~> VBool(v)));
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun addEvent(name: String, attributes: Map<String, Value>): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    attributesCopy = mutable Map[];
+    attributes.each((k, v) ->
+      attributesCopy.set(SKStore.duplicate(k), v.duplicate())
+    );
+    span.addEvent(
+      Event::create(
+        SKStore.duplicate(name),
+        unsafe_chill_trust_me(attributesCopy),
+      ),
+    );
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun setOkStatus(): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setStatus(SOk());
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun setErrorStatus(description: String = ""): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setStatus(SError(), SKStore.duplicate(description));
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun setUnsetStatus(): void {
+  get_span().each(span -> {
+    SKStore.switchToTrace();
+    span.setStatus(SUnset());
+    SKStore.restoreFromTrace()
+  })
+}
+
+@debug
+fun endSpan(): void {
+  monitoring() match {
+  | MOn _ ->
+    get_span().each(span -> {
+      SKStore.switchToTrace();
+      span.!optEnd = Some(now());
+      if (span.optParent.isNone()) {
+        memory = memory();
+        span.setIntAttribute("skruntime.memory.persistent", memory.persistent);
+        span.setIntAttribute("skruntime.memory.freetable", memory.freetable);
+        span.setIntAttribute("skruntime.obstack.pick", memory.obstack_pick);
+      };
+      impact = get_trace_size(span.optParent.map(p -> p.i1));
+      span.setIntAttribute("skmonitor.memory", impact);
+      flush(span);
```

Comment:
So this is when we write information out, at the grain of a span. Totally reasonable. But, do I understand correctly: we learn about events when the span ends, and we'll get a parent id, but we won't learn any info about that parent until its span ends? For a long lived process like write-csv if the parent span lasts the whole time of the process and contains e.g. the user we won't be able to figure out who's affected until that process ends?

This all seems reasonable, I just wanted to check, and to point out we need to be careful about what info we store where and when we'll get it.

Reply:
You understand correctly. I had the same thought when I looked at OpenTelemetry.
Like in OpenTelemetry Protocol File Exporter.
- There is no guarantee that the data is ordered.
- There is no guarantee in particular that timestamps will be monotonically increasing.
Some data could be added the children to identify more precisely the parent.

If we need to have some kind validation on span tree, it could be done in the "Collector"